synapse/lib/modelrev.py

```diff
@@ -783,6 +784,292 @@ async def revModel_0_2_26(self, layers):
+
+        for $view in $lib.view.list(deporder=$lib.true) {
+
+            if (not $layers.has($view.layers.0.iden)) { continue }
+
+            view.exec $view.iden {
+
+                $layer = $lib.layer.get()
+
+                for $oldcpe in $layer.liftByProp(it:sec:cpe) {
+                    $info = $lib.model.migration.s.itSecCpe_2_170_0_internal($oldcpe)
+
+                    $repr = $oldcpe.repr()
+
+                    $success = ($info.status = "success")
+
+                    if $success {
+                        // No primary property changes, nothing to do. Node has been fully migrated.
+                        if (not $info.valu) {
+                            continue
+                        }
+
+                        $lib.log.debug(`Migrating invalid it:sec:cpe node: {$repr}`)
+
+                        /*
+                         * At this point, we have a node that can be fixed but
+                         * needs to be migrated to a new node because the primary
+                         * property needs to be changed. We'll create a new
+                         * (correct) node, and copy everything from the old node.
+                         * Then we complete the migration by iterating through
+                         * all the views to fix the references.
+                         */
+
+                        // Create the new CPE node and migrate the edges, tags, and data
+                        $newcpe = {
+                            [ it:sec:cpe=$info.valu ]
+                            $lib.model.migration.copyEdges($oldcpe, $node)
+                            $lib.model.migration.copyTags($oldcpe, $node)
+                            $lib.model.migration.copyData($oldcpe, $node)
+                        }
+
+                        // Iterate through the views again and fix up all the references
+                        for $rview in $lib.view.list(deporder=$lib.true) {
+                            if (not $layers.has($rview.layers.0.iden)) { continue }
+
+                            view.exec $rview.iden {
+
+                                // Fix references that point to old node to now point to new node
+                                for ($form, $prop, $isarray) in $refforms {
+                                    $n = $lib.model.migration.liftByPropValuNoNorm($form, $prop, $repr)
+                                    if (not $n) { continue }
+
+                                    yield $n
+
+                                    if $isarray {
+
+                                        // We can't just [ :$prop-=$repr ] because the norm() function gets called
+                                        // on the array type deep down in the AST. So, instead, we have to operate on
+                                        // the whole array.
+
+                                        $valu = $lib.copy(:$prop)
+                                        $valu.remove($repr)
```

Comment:
Does this handle the case where a non-uniq array has multiple of the same invalid CPE?